lms/djangoapps/completion_api/views.py

```diff
@@ -254,9 +259,21 @@ def get(self, request):
         Handler for GET requests.
         """
         self.paginator = self.pagination_class()  # pylint: disable=attribute-defined-outside-init
+        mobile_only = (self.request.query_params.get('mobile_only', 'false')).lower() == 'true'
 
         # Paginate the list of active enrollments, annotated (manually) with a student progress object.
-        enrollments = CourseEnrollment.objects.filter(user=self.get_user(), is_active=True).order_by('course_id')
+        enrollments = CourseEnrollment.objects.filter(
+            user=self.get_user(),
+            is_active=True).order_by('course_id')
+
+        if mobile_only:
+            course_keys = []
+            for course_enrollment in enrollments:
+                course_keys.append(course_enrollment.course_id)
+            course_overview_list = CourseOverview.objects.filter(id__in=course_keys, mobile_available=True)
+            filtered_course_overview = [overview.id for overview in course_overview_list]
+            enrollments = enrollments.filter(course_id__in=filtered_course_overview)
```

Comment:
This seems a little excessive.  Get the enrollments to get the overviews, then get the enrollments again.  Can you join to the course overview table and filter in the first query, so we only have to hit the DB once instead of three times?

Maybe using [Subselects with values](https://docs.djangoproject.com/en/1.11/ref/models/querysets/#in).  Something like:

    enrollments = CourseEnrollments.objects.filter(
        user=self.get_user(),
        is_active=True,
        course_id__in=CourseOverview.objects.filter(mobile_available=True).values('course_id')
    )

Reply:
This doesn't work because the call to values turns the ID objects into strings which causes an exception. I'm looking at another way. Measuring the queries says there are two queries rather than 3.

Reply:
Oh.  That's frustrating about the course_ids.

Re: 2 queries vs 3, that seems odd because there were very clearly three requests made.  Maybe where you were measuring it the last one hadn't been evaluated yet? (Django only evaluates querysets (sends the request to the database) when you call .all(), .count(), or .exists(), or iterate over the queryset).

The revised version unfortunately isn't able to use the queryset as a subquery, because we pull the data into python to create the list of course_keys. The CourseOverview query is now evaluated against all courses in the catalog.  Given the constraint that the CourseKeyField doesn't seem to work with subqueries, I think your original version is actually better.  There's one extra query, but it's bounded to only retrieve courses the user is interested in, so there should be significantly less network traffic generated between the app server and the database. 

Other than that, everything looks great on my end.  Pending a revert of the latest change to the query, and assuming tests pass, this looks good to go, :+1:.